syntax = "proto2";

package gpu.cldrive;

import "gpu/clinfo/proto/clinfo.proto";
import "gpu/libcecl/proto/libcecl.proto";

message CldriveInstance {
  required gpu.clinfo.OpenClDevice device = 1;
  required string opencl_src = 2;
  optional DynamicParams dynamic_params = 3;
  required int32 min_runs_per_kernel = 4;
  // Output fields:

  enum InstanceOutcome {
    UNKNOWN_ERROR = 0; // Default value is an error.
    PASS = 1;
    PROGRAM_COMPILATION_FAILURE = 2;
    NO_KERNELS_IN_PROGRAM = 3;
  }
  optional InstanceOutcome outcome = 5;
  optional CldriveKernelInstance kernel = 6;
}

message CldriveKernelInstance {
  optional CldriveKernelRun run = 1;
  // Per-work-item memory requirements of the kernel.
  optional string name = 2;
  enum KernelInstanceOutcome {
    UNKNOWN_ERROR = 0; // Default value is an error.
    PASS = 1;
    NO_ARGUMENTS = 2;
    NO_MUTABLE_ARGUMENTS = 3;
    UNSUPPORTED_ARGUMENTS = 4;
  }
  optional KernelInstanceOutcome outcome = 3;
  optional int64 work_item_local_mem_size_in_bytes = 4;
  optional int64 work_item_private_mem_size_in_bytes = 5;
}

message DynamicParams {
  required int32 global_size_x = 1;
  required int32 local_size_x = 2;
}

message CldriveKernelRun {
  optional gpu.libcecl.OpenClKernelInvocation log = 1;
  required KernelRunOutcome outcome = 2;
  enum KernelRunOutcome {
    UNKNOWN_ERROR = 0;
    PASS = 1;
    INVALID_KERNEL = 2;
    INVALID_DYNAMIC_PARAMS = 4;
    NO_OUTPUT = 5;
    INPUT_INSENSITIVE = 6;
    NONDETERMINISTIC = 7;
  }
}
